File: code/cpp/ch13/practice/warmup_02_generate_subsets.cpp
/*
 * Warmup 2: Generate All Subsets
 * ====================================
 * Chapter 13: Bronze Battle Plan — Complete Search & Simulation
 *
 * PROBLEM:
 *   Given a vector of distinct integers, return all subsets
 *   sorted by size then lexicographically.
 *
 * EXAMPLES:
 *   solve({1,2,3}) -> {{},{1},{2},{3},{1,2},{1,3},{2,3},{1,2,3}}
 *   solve({0})     -> {{},{0}}
 *
 * CONSTRAINTS:
 *   - 0 <= nums.size() <= 10
 *   - All elements are distinct
 */

#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

vector<vector<int>> solve(vector<int> nums) {
    // TODO: Replace this with your solution
    return {};
}

// -- Do not change anything below this line --------------------------
int main() {
    int n;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++) cin >> nums[i];
    vector<vector<int>> result = solve(nums);
    for (auto& s : result) {
        for (int i = 0; i < (int)s.size(); i++) {
            if (i > 0) cout << " ";
            cout << s[i];
        }
        cout << endl;
    }
    return 0;
}

File: code/python/ch13/practice/warmup_02_generate_subsets.py
"""
Warmup 2: Generate All Subsets
================================
Chapter 13: Bronze Battle Plan — Putting It All Together

PROBLEM
-------
Given a list of distinct integers, return all possible subsets sorted
by length first, then lexicographically within each length group.

INPUT FORMAT
------------
A single line of space-separated integers (may be empty).

OUTPUT FORMAT
-------------
Each subset on its own line as a list.

CONSTRAINTS
-----------
- 0 <= len(nums) <= 10
- -100 <= nums[i] <= 100
- All elements are distinct

EXAMPLES
--------
Input:
  1 2 3
Output:
  []
  [1]
  [2]
  [3]
  [1, 2]
  [1, 3]
  [2, 3]
  [1, 2, 3]

Input:
  (empty)
Output:
  []

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
The main block at the bottom handles input/output — don't change it.
"""


def solve(nums: list[int]) -> list[list[int]]:
    """Return all subsets sorted by length then lexicographically."""
    pass  # TODO: Replace this with your solution



# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    line = input().strip()
    if line:
        nums = list(map(int, line.split()))
    else:
        nums = []
    result = solve(nums)
    for s in result:
        print(s)

